detect_test.go

```diff
@@ -93,33 +67,95 @@ func testDetect(t *testing.T, context spec.G, it spec.S) {
 							Name: poetry.CPython,
 							Metadata: poetry.BuildPlanMetadata{
 								Build:         true,
-								Version:       "9.8.7",
+								Version:       "1.2.3",
 								VersionSource: "pyproject.toml",
 							},
 						},
-						{
-							Name: "poetry",
-							Metadata: poetry.BuildPlanMetadata{
-								VersionSource: "BP_POETRY_VERSION",
-								Version:       "some-version",
-							},
-						},
 					},
 				},
 			}))
 		})
-	}, spec.Sequential())
 
-	context("error handling", func() {
-		it("handles an error from the pyproject.toml parser", func() {
-			expectedErr := errors.New("hi")
-			parsePythonVersion.ParsePythonVersionCall.Returns.Error = expectedErr
+		context("when the BP_POETRY_VERSION is set", func() {
+			it.Before(func() {
+				Expect(os.Setenv("BP_POETRY_VERSION", "some-version")).To(Succeed())
+			})
 
-			result, err := detect(packit.DetectContext{
-				WorkingDir: "/working-dir",
+			it.After(func() {
+				Expect(os.Unsetenv("BP_POETRY_VERSION")).To(Succeed())
+			})
+
+			it("returns a plan that requires that version of poetry", func() {
+				parsePythonVersion.ParsePythonVersionCall.Returns.String = "9.8.7"
+
+				result, err := detect(packit.DetectContext{
+					WorkingDir: workingDir,
+				})
+				Expect(err).NotTo(HaveOccurred())
+				Expect(parsePythonVersion.ParsePythonVersionCall.Receives.String).To(Equal(filepath.Join(workingDir, "pyproject.toml")))
+				Expect(result).To(Equal(packit.DetectResult{
+					Plan: packit.BuildPlan{
+						Provides: []packit.BuildPlanProvision{
+							{Name: "poetry"},
+						},
+						Requires: []packit.BuildPlanRequirement{
+							{
+								Name: poetry.Pip,
+								Metadata: poetry.BuildPlanMetadata{
+									Build: true,
+								},
+							},
+							{
+								Name: poetry.CPython,
+								Metadata: poetry.BuildPlanMetadata{
+									Build:         true,
+									Version:       "9.8.7",
+									VersionSource: "pyproject.toml",
+								},
+							},
+							{
+								Name: "poetry",
+								Metadata: poetry.BuildPlanMetadata{
+									VersionSource: "BP_POETRY_VERSION",
+									Version:       "some-version",
+								},
+							},
+						},
+					},
+				}))
+			})
+		})
+
+		context("error handling", func() {
+			// the python dependency is mandatory
+			// https://python-poetry.org/docs/pyproject/#dependencies-and-dev-dependencies
+			it("fails detection when no python version found", func() {
```

Comment:
The pattern I've seen for other buildpacks is to consider correctly-failing detection as a happy-path rather than an error case. So we would move it up to the section above rather than having it under the error handling context.

Reply:
As discussed, we can modify this at a later date.